server/src/services/stats.ts

```diff
@@ -101,4 +107,71 @@ export class StatsService extends BaseService {
       withoutBots: await this.getLeaderboardWithoutBots(limit),
     };
   }
+
+  async updateSoloHighScore(gameId: number, points: number, maxRound: number): Promise<void> {
+    // update the solo highscores table with data from a victory game and return the player's entry with rank
+    const highscoreRepo = this.em.getRepository(SoloHighScore);
+    const pointsPerRound = points / maxRound;
+    const game = await this.em.getRepository(SoloGame).findOneOrFail(gameId, {
+      relations: ["player", "player.user"],
+    });
+    const user = game.player.user;
+    let highscore = await highscoreRepo.findOne({ user });
+
+    if (highscore) {
+      if (pointsPerRound > highscore.pointsPerRound) {
+        highscore.pointsPerRound = pointsPerRound;
+        highscore.maxRound = maxRound;
+        highscore.points = points;
+        highscore.gameId = gameId;
+        await highscoreRepo.save(highscore);
+      }
+    } else {
+      const newHighscore = highscoreRepo.create({
+        user,
+        maxRound,
+        pointsPerRound,
+        points,
+        gameId,
+      });
+      await highscoreRepo.save(newHighscore);
+      highscore = newHighscore;
+    }
+  }
+
+  async getSoloHighScoreData(limit: number): Promise<SoloHighScoreData> {
+    // get the top players in the solo highscores table
+    const highscoresData = await this.em
+      .getRepository(SoloHighScore)
+      .createQueryBuilder("highscores")
+      .select("ROW_NUMBER() OVER (ORDER BY highscores.pointsPerRound DESC)::integer as rank")
```

Comment:
this SQL query would benefit from more detailed comments @virtualcommons/port-of-mars-dev